### PR TITLE
refactor: 6.0 mparticle

### PR DESCRIPTION
**Pull request by PoornimaApptentive**
Tested the all the interactions except rating dialog. Tested the Person data and Custom Person data.

Code change: Replaced the AttributeListener with UserAttributeLIstener & IdentityListener. Updating the mParticle Id on Apptentive registration success & OnIdentityUpdated (//TODO need to see if the update is needed here too)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -43,6 +46,37 @@ android {
     }
 }
```
Comment (PoornimaApptentive) [inline build.gradle:48]: Temporary code for testing.

File: src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt [hunk 2]
```diff
@@ -41,88 +36,98 @@ class ApptentiveKit : KitIntegration(), KitIntegration.EventListener, CommerceLi
...
+    }
 
-    override fun setAllUserAttributes(
-        attributes: Map<String, String>,
-        attributeLists: Map<String, List<String>>
+    override fun onSetUserAttributeList(
+        attributeKey: String?,
+        attributeValueList: MutableList<String>?,
+        user: FilteredMParticleUser?
     ) {
```
Comment (PoornimaApptentive) [inline src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt:106]: //TODO

File: build.gradle [hunk 3]
```diff
@@ -20,6 +22,7 @@ buildscript {
 plugins {
     id "org.sonarqube" version "3.5.0.2730"
     id "org.jlleitschuh.gradle.ktlint" version "11.1.0"
```
Comment (PoornimaApptentive) [inline build.gradle:24]: Temporary code for testing

File: src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt [hunk 4]
```diff
@@ -41,88 +40,100 @@ class ApptentiveKit : KitIntegration(), KitIntegration.EventListener, CommerceLi
 
         if (apptentiveAppKey != null && apptentiveAppSignature != null) {
             val configuration = ApptentiveConfiguration(apptentiveAppKey, apptentiveAppSignature)
-            Apptentive.register(context.applicationContext as Application, configuration)
-            ApptentiveNotificationCenter.defaultCenter()
-                .addObserver(
-                    ApptentiveNotifications.NOTIFICATION_CONVERSATION_STATE_DID_CHANGE,
-                    this
-                )
+            configuration.logLevel = LogLevel.Verbose
```
Comment (PoornimaApptentive) [inline src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt:43]: Temporarily set them the logLevel and shouldSanitizeLogMessages here.

File: src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt [hunk 5]
```diff
@@ -159,129 +170,61 @@ class ApptentiveKit : KitIntegration(), KitIntegration.EventListener, CommerceLi
...
+        user?.userIdentities?.entries?.forEach {
+            when (it.key) {
+                IdentityType.CustomerId ->  {
+                    if (KitUtils.isEmpty(Apptentive.getPersonName())) {
+                        // Use id as customer name iff no full name is set yet.
+                        Apptentive.setPersonName(it.value)
+                    }
+                }
+                IdentityType.Email -> Apptentive.setPersonEmail(it.value)
+                else -> Log.d("UserIdentity", "Other type")
```
Comment (ChaseApptentive) [inline src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt:221]: Can we use the Apptentive version of Log throughout the ApptentiveKit?

Something like
`Log.d(LogTag("mParticle"), "User Identity: Other type")`

File: src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt [hunk 6]
```diff
@@ -2,33 +2,32 @@ package com.mparticle.kits
...
+class ApptentiveKit : KitIntegration(), KitIntegration.EventListener, IdentityListener,
+    UserAttributeListener  {
     private var enableTypeDetection = false
     private var lastKnownFirstName: String? = null
     private var lastKnownLastName: String? = null
+
+    //region KitIntegration
     override fun getName(): String = NAME
+    @OptIn(InternalUseOnly::class)
     override fun onKitCreate(
```
Comment (ChaseApptentive) [inline src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt:31]: Could we add a line break after the getName method

File: src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt [hunk 7]
```diff
@@ -159,129 +170,61 @@ class ApptentiveKit : KitIntegration(), KitIntegration.EventListener, CommerceLi
...
-        event: String,
-        customData: Map<String, String>?,
-        vararg extendedData: ExtendedData
-    ) {
-        Apptentive.engage(context, event, parseCustomData(customData), *extendedData)
+    private fun engage(event: String, customData: Map<String, String>?) {
+        Apptentive.engage(event, parseCustomData(customData))
     }
 
     /* Apptentive SDK does not provide a function which accepts Object as custom data so we need to cast */
```
Comment (ChaseApptentive) [inline src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt:230]: @frankus Just want to check that iOS is doing the same with `addCustomPersonData` and `addCustomDeviceData`. We always add a string version of the custom data and if it's a Boolean or Number then we add an additional custom data with the type appended to the end.

File: src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt [hunk 8]
```diff
@@ -41,88 +40,100 @@ class ApptentiveKit : KitIntegration(), KitIntegration.EventListener, CommerceLi
...
+                    lastKnownLastName = value.toString()
+                }
+                else -> {
+                    addCustomPersonData(key, value.toString())
+                    return
+                }
             }
-            fullName += lastKnownLastName
+            val fullName = listOfNotNull(lastKnownLastName, lastKnownFirstName).joinToString(separator = " ")
+            if (fullName.isNotBlank()) Apptentive.setPersonName(fullName.trim())
```
Comment (ChaseApptentive) [inline src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt:100]: Looks like this is doing Last name + First name. Is that the correct order?
Could we also put the `trim()` function after `joinToString`?

File: src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt [hunk 9]
```diff
@@ -41,88 +40,100 @@ class ApptentiveKit : KitIntegration(), KitIntegration.EventListener, CommerceLi
...
-                firstName = value
-            } else if (key.equals(MParticle.UserAttributes.LASTNAME, true)) {
-                lastName = value
-            } else {
-                addCustomPersonData(key, value)
+        userAttributes?.let { userAttribute ->
+            val firstName = userAttribute[MParticle.UserAttributes.FIRSTNAME] ?: ""
+            val lastName = userAttribute[MParticle.UserAttributes.LASTNAME] ?: ""
+            val fullName = listOfNotNull(firstName, lastName).joinToString(separator = " ")
+            if (fullName.isNotBlank()) Apptentive.setPersonName(fullName.trim())
```
Comment (ChaseApptentive) [inline src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt:125]: This ordering looks correct, but could we move the `trim` here as well? Unless we have a reason to not move it up.

File: src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt [hunk 10]
```diff
@@ -159,129 +170,61 @@ class ApptentiveKit : KitIntegration(), KitIntegration.EventListener, CommerceLi
...
+
     //region Helpers
-    private fun engage(context: Context, event: String, customData: Map<String, String>?) {
-        engage(context, event, customData, *arrayOf())
+    private fun setUserIdentity(user: MParticleUser?) {
+        user?.userIdentities?.entries?.forEach {
+            when (it.key) {
+                IdentityType.CustomerId ->  {
+                    if (KitUtils.isEmpty(Apptentive.getPersonName())) {
+                        // Use id as customer name iff no full name is set yet.
```
Comment (ChaseApptentive) [inline src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt:216]: Typo here `iff`

File: src/main/kotlin/com/mparticle/kits/StringUtils.kt [hunk 11]
```diff
@@ -11,52 +8,33 @@ internal object StringUtils {
...
     }
 
     @JvmStatic
     fun tryParseNumber(value: String): Number? {
         val longValue = tryParseLong(value)
         return if (longValue != null) {
-            if (longValue >= Int.MIN_VALUE && longValue <= Int.MAX_VALUE) {
-                longValue.toInt()
-            } else longValue
+            if (isInIntegerRange(longValue)) longValue.toInt()
```
Comment (ChaseApptentive) [inline src/main/kotlin/com/mparticle/kits/StringUtils.kt:18]: Does this need to be converted to an Int for any reason?

File: src/main/kotlin/com/mparticle/kits/StringUtils.kt [hunk 12]
```diff
@@ -11,52 +8,33 @@ internal object StringUtils {
...
             value.toLong()
         } catch (e: NumberFormatException) {
             null
         }
     }
 
-    fun tryParseDouble(value: String): Double? {
+    private fun tryParseDouble(value: String): Double? {
         return try {
             java.lang.Double.valueOf(value)
```
Comment (ChaseApptentive) [inline src/main/kotlin/com/mparticle/kits/StringUtils.kt:35]: I think there's a Kotlin version of this. `value.ToDouble` or something

File: src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt (hunk 5 above)
Comment (PoornimaApptentive) [reply src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt:221]: I had to use the Android Util as Apptentive Logs and LogTags are internal to the SDKs.

File: src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt (hunk 8 above)
Comment (PoornimaApptentive) [reply src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt:100]: Another one? I would move it instead.

Comment (PoornimaApptentive) [reply src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt:100]: Legacy has the Last name + First name pattern. To keep it consistent for the migration left it that way.

File: src/main/kotlin/com/mparticle/kits/StringUtils.kt (hunk 11 above)
Comment (PoornimaApptentive) [reply src/main/kotlin/com/mparticle/kits/StringUtils.kt:18]: I have no idea.

File: src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt (hunk 5 above)
Comment (ChaseApptentive) [reply src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt]: That should be okay though right, We can use the `@OptIn(InternalUseOnly::class)` and it should be fine.

Comment (PoornimaApptentive) [reply src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt]: Yeah, forgot to update the comment. The latest commit uses the Apptentive log.

Comment (PoornimaApptentive) [reply src/main/kotlin/com/mparticle/kits/ApptentiveKit.kt]: Refer this PR https://github.com/apptentive/mparticle-android-integration-apptentive/pull/21/files#diff-8f5282fb1218479826343546221395be574ff015f9ad20973a658e17a183ed16

I accidentally closed this one and couldn't open it again.

**Pull request closed without merging**